File: csharp/v7/Dart/CSharpLexerBase.dart
import 'dart:io';
import 'package:antlr4/antlr4.dart';
import 'CSharpLexer.dart';

abstract class CSharpLexerBase extends Lexer {
    int interpolatedStringLevel = 0;
    List<bool> interpolatedVerbatiums = [];
    List<int> curlyLevels = [];
    bool verbatium = false;

    // Preprocessor state
    final List<Token> _pending = [];
    final Set<String> _symbols = {};
    final List<bool> _condition = [];
    final List<bool> _taken = [];

    // Expression evaluator cursor
    List<Token> _expr = [];
    int _epos = 0;

    CSharpLexerBase(CharStream input) : super(input) {
        _initPreprocessor();
    }

    void OnInterpolatedRegularStringStart() {
        interpolatedStringLevel++;
        interpolatedVerbatiums.add(false);
        verbatium = false;
    }

    void OnInterpolatedVerbatiumStringStart() {
        interpolatedStringLevel++;
        interpolatedVerbatiums.add(true);
        verbatium = true;
    }

    void OnOpenBrace() {
        if (interpolatedStringLevel > 0) {
            curlyLevels[curlyLevels.length - 1]++;
        }
    }

    void OnCloseBrace() {
        if (interpolatedStringLevel > 0) {
            curlyLevels[curlyLevels.length - 1]--;
            if (curlyLevels.last == 0) {
                curlyLevels.removeLast();
                skip();
                popMode();
            }
        }
    }

    void OnColon() {
        if (interpolatedStringLevel > 0) {
            var ind = 1;
            var switchToFormatString = true;
            while ((inputStream.LA(ind) ?? -1) != 0x7D /* } */) {
                final ch = inputStream.LA(ind) ?? -1;
                if (ch == 0x3A /* : */ || ch == 0x29 /* ) */) {
                    switchToFormatString = false;
                    break;
                }
                ind++;
            }
            if (switchToFormatString) {
                mode(CSharpLexer.INTERPOLATION_FORMAT);
            }
        }
    }

    void OpenBraceInside() {
        curlyLevels.add(1);
    }

    void OnDoubleQuoteInside() {
        interpolatedStringLevel--;
        interpolatedVerbatiums.removeLast();
        verbatium = interpolatedVerbatiums.isNotEmpty && interpolatedVerbatiums.last;
    }

    void OnCloseBraceInside() {
        curlyLevels.removeLast();
    }

    bool IsRegularCharInside() {
        return !verbatium;
    }

    bool IsVerbatiumDoubleQuoteInside() {
        return verbatium;
    }

    // -------------------------------------------------------------------------
    // Preprocessor initialisation
    // -------------------------------------------------------------------------
    void _initPreprocessor() {
        for (final arg in Platform.executableArguments) {
            if (arg.startsWith("--D")) {
                for (final sym in arg.substring(3).split(';')) {
                    if (sym.isNotEmpty) _symbols.add(sym);
                }
            }
        }
    }

    bool _isActive() {
        return _condition.isEmpty || _condition.last;
    }

    // -------------------------------------------------------------------------
    // nextToken override — intercepts DIRECTIVE-channel tokens
    // -------------------------------------------------------------------------
    @override
    Token nextToken() {
        if (_pending.isNotEmpty) return _pending.removeAt(0);

        final tok = super.nextToken();

        if (tok.channel == CSharpLexer.DIRECTIVE) {
            Token? skipped;
            switch (tok.type) {
                case CSharpLexer.TOKEN_DEFINE: _handleDefine(); break;
                case CSharpLexer.TOKEN_UNDEF:  _handleUndef();  break;
                case CSharpLexer.TOKEN_IF:     skipped = _handleIf();   break;
                case CSharpLexer.TOKEN_ELIF:   skipped = _handleElif(); break;
                case CSharpLexer.TOKEN_ELSE:   skipped = _handleElse(); break;
                case CSharpLexer.TOKEN_ENDIF:  _handleEndif();  break;
            }
            if (skipped != null) _pending.add(skipped);
        }

        return tok;
    }

    // -------------------------------------------------------------------------
    // Directive handlers
    // -------------------------------------------------------------------------
    void _handleDefine() {
        final line = _collectLine();
        final sym = _symbolFromLine(line);
        if (_isActive() && sym != null) _symbols.add(sym);
    }

    void _handleUndef() {
        final line = _collectLine();
        final sym = _symbolFromLine(line);
        if (_isActive() && sym != null) _symbols.remove(sym);
    }

    Token? _handleIf() {
        final line = _collectLine();
        final outer = _isActive();
        final result = outer && _evaluate(line);
        _condition.add(result);
        _taken.add(result);
        return result ? null : _skipFalseBlock();
    }

    Token? _handleElif() {
        final line = _collectLine();
        final alreadyTaken = _taken.isNotEmpty ? _taken.removeLast() : false;
        if (_condition.isNotEmpty) _condition.removeLast();
        final outer = _isActive();
        final result = !alreadyTaken && outer && _evaluate(line);
        _condition.add(result);
        _taken.add(alreadyTaken || result);
        return result ? null : _skipFalseBlock();
    }

    Token? _handleElse() {
        _collectLine();
        final alreadyTaken = _taken.isNotEmpty ? _taken.removeLast() : false;
        if (_condition.isNotEmpty) _condition.removeLast();
        final outer = _isActive();
        final result = !alreadyTaken && outer;
        _condition.add(result);
        _taken.add(true);
        return result ? null : _skipFalseBlock();
    }

    void _handleEndif() {
        _collectLine();
        if (_condition.isNotEmpty) _condition.removeLast();
        if (_taken.isNotEmpty)    _taken.removeLast();
    }

    // -------------------------------------------------------------------------
    // _collectLine — drain DIRECTIVE_MODE tokens up to DIRECTIVE_NEW_LINE
    // -------------------------------------------------------------------------
    List<Token> _collectLine() {
        final tokens = <Token>[];
        Token t;
        do {
            t = super.nextToken();
            if (t.channel != Token.HIDDEN_CHANNEL
                && t.channel != CSharpLexer.COMMENTS_CHANNEL)
                tokens.add(t);
        } while (t.type != CSharpLexer.TOKEN_DIRECTIVE_NEW_LINE
              && t.type != Token.EOF);
        return tokens;
    }

    String? _symbolFromLine(List<Token> line) {
        for (final t in line)
            if (t.type == CSharpLexer.TOKEN_CONDITIONAL_SYMBOL) return t.text;
        return null;
    }

    // -------------------------------------------------------------------------
    // _skipFalseBlock — scan char stream, return SKIPPED_SECTION on HIDDEN channel
    // -------------------------------------------------------------------------
    Token _skipFalseBlock() {
        final buf = StringBuffer();
        final stream = inputStream;
        var depth = 1;
        var atLineStart = true;
        final startLine = line;

        while (true) {
            final c = stream.LA(1) ?? Token.EOF;
            if (c == Token.EOF) break;

            if (c == 0x0D || c == 0x0A || c == 0x85 || c == 0x2028 || c == 0x2029) {
                stream.consume();
                buf.writeCharCode(c);
                if (c == 0x0D && (stream.LA(1) ?? -1) == 0x0A) {
                    stream.consume();
                    buf.write('\n');
                }
                atLineStart = true;
                continue;
            }

            if (atLineStart && (c == 0x20 || c == 0x09)) {
                stream.consume();
                buf.writeCharCode(c);
                continue;
            }

            if (atLineStart && c == 0x23 /* '#' */) {
                final kw = _peekKeyword();
                if (kw == "if") {
                    depth++;
                } else if (kw == "endif") {
                    depth--;
                    if (depth == 0) break;
                } else if ((kw == "else" || kw == "elif") && depth == 1) {
                    break;
                }
            }

            atLineStart = false;
            stream.consume();
            buf.writeCharCode(c);
        }

        final tok = CommonTokenFactory.DEFAULT.create(
            CSharpLexer.TOKEN_SKIPPED_SECTION,
            buf.toString(),
            Pair(this, inputStream),
            Token.HIDDEN_CHANNEL,
            -1, -1,
            startLine, 0);
        return tok;
    }

    String _peekKeyword() {
        var i = 2; // LA(1) is '#'
        while (inputStream.LA(i) == 0x20 || inputStream.LA(i) == 0x09) i++;
        final buf = StringBuffer();
        int c;
        while (true) {
            final cv = inputStream.LA(i);
            if (cv == null || cv == -1) break;
            c = cv;
            if ((c < 0x61 || c > 0x7A) && (c < 0x41 || c > 0x5A)) break;
            buf.writeCharCode(c);
            i++;
        }
        return buf.toString();
    }

    // -------------------------------------------------------------------------
    // Recursive-descent expression evaluator
    // -------------------------------------------------------------------------
    bool _evaluate(List<Token> tokens) {
        _expr = tokens;
        _epos = 0;
        return _parseOr();
    }

    int _peekType() {
        if (_epos < _expr.length) {
            final t = _expr[_epos].type;
            if (t != CSharpLexer.TOKEN_DIRECTIVE_NEW_LINE && t != Token.EOF) return t;
        }
        return -1;
    }

    Token _eConsume() => _expr[_epos++];

    bool _parseOr() {
        var v = _parseAnd();
        while (_peekType() == CSharpLexer.TOKEN_OP_OR) { _eConsume(); v = _parseAnd() || v; }
        return v;
    }

    bool _parseAnd() {
        var v = _parseEq();
        while (_peekType() == CSharpLexer.TOKEN_OP_AND) { _eConsume(); v = _parseEq() && v; }
        return v;
    }

    bool _parseEq() {
        final v = _parseUnary();
        if (_peekType() == CSharpLexer.TOKEN_OP_EQ) { _eConsume(); return v == _parseUnary(); }
        if (_peekType() == CSharpLexer.TOKEN_OP_NE) { _eConsume(); return v != _parseUnary(); }
        return v;
    }

    bool _parseUnary() {
        if (_peekType() == CSharpLexer.TOKEN_BANG) { _eConsume(); return !_parseUnary(); }
        return _parsePrimary();
    }

    bool _parsePrimary() {
        final t = _peekType();
        if (t == CSharpLexer.TOKEN_TRUE)               { _eConsume(); return true; }
        if (t == CSharpLexer.TOKEN_FALSE)              { _eConsume(); return false; }
        if (t == CSharpLexer.TOKEN_CONDITIONAL_SYMBOL) { return _symbols.contains(_eConsume().text ?? ""); }
        if (t == CSharpLexer.TOKEN_OPEN_PARENS) {
            _eConsume();
            final v = _parseOr();
            if (_peekType() == CSharpLexer.TOKEN_CLOSE_PARENS) _eConsume();
            return v;
        }
        return false;
    }
}

File: csharp/v7/Cpp/CSharpLexerBase.cpp
#include "antlr4-runtime.h"
#include "CSharpLexerBase.h"
#include "CSharpLexer.h"
#include <cctype>

CSharpLexerBase::CSharpLexerBase(antlr4::CharStream *input)
    : antlr4::Lexer(input)
{
}

// -------------------------------------------------------------------------
// Interpolated string helpers
// -------------------------------------------------------------------------
void CSharpLexerBase::OnInterpolatedRegularStringStart()
{
    interpolatedStringLevel++;
    interpolatedVerbatiums.push_back(false);
    verbatium = false;
}

void CSharpLexerBase::OnInterpolatedVerbatiumStringStart()
{
    interpolatedStringLevel++;
    interpolatedVerbatiums.push_back(true);
    verbatium = true;
}

void CSharpLexerBase::OnOpenBrace()
{
    if (interpolatedStringLevel > 0)
        curlyLevels.back()++;
}

void CSharpLexerBase::OnCloseBrace()
{
    if (interpolatedStringLevel > 0)
    {
        curlyLevels.back()--;
        if (curlyLevels.back() == 0)
        {
            curlyLevels.pop_back();
            skip();
            popMode();
        }
    }
}

void CSharpLexerBase::OnColon()
{
    if (interpolatedStringLevel > 0)
    {
        int ind = 1;
        bool switchToFormatString = true;
        auto *stream = dynamic_cast<antlr4::CharStream *>(getInputStream());
        while (stream->LA(ind) != '}')
        {
            auto ch = stream->LA(ind);
            if (ch == ':' || ch == ')')
            {
                switchToFormatString = false;
                break;
            }
            ind++;
        }
        if (switchToFormatString)
            setMode(CSharpLexer::INTERPOLATION_FORMAT);
    }
}

void CSharpLexerBase::OpenBraceInside()
{
    curlyLevels.push_back(1);
}

void CSharpLexerBase::OnDoubleQuoteInside()
{
    interpolatedStringLevel--;
    if (!interpolatedVerbatiums.empty())
        interpolatedVerbatiums.pop_back();
    verbatium = !interpolatedVerbatiums.empty() && interpolatedVerbatiums.back();
}

void CSharpLexerBase::OnCloseBraceInside()
{
    if (!curlyLevels.empty())
        curlyLevels.pop_back();
}

bool CSharpLexerBase::IsRegularCharInside() { return !verbatium; }
bool CSharpLexerBase::IsVerbatiumDoubleQuoteInside() { return verbatium; }

bool CSharpLexerBase::isActive() const
{
    return condition_.empty() || condition_.back();
}

// -------------------------------------------------------------------------
// nextToken override — intercepts DIRECTIVE-channel tokens
// -------------------------------------------------------------------------
std::unique_ptr<antlr4::Token> CSharpLexerBase::nextToken()
{
    if (!pending_.empty())
    {
        auto tok = std::move(pending_.front());
        pending_.pop_front();
        return tok;
    }

    auto tok = antlr4::Lexer::nextToken();

    if (tok->getChannel() == CSharpLexer::DIRECTIVE)
    {
        std::unique_ptr<antlr4::Token> skipped;
        switch (tok->getType())
        {
        case CSharpLexer::DEFINE: handleDefine(); break;
        case CSharpLexer::UNDEF:  handleUndef();  break;
        case CSharpLexer::IF:     skipped = handleIf();   break;
        case CSharpLexer::ELIF:   skipped = handleElif(); break;
        case CSharpLexer::ELSE:   skipped = handleElse(); break;
        case CSharpLexer::ENDIF:  handleEndif();  break;
        }
        if (skipped)
            pending_.push_back(std::move(skipped));
    }

    return tok;
}

// -------------------------------------------------------------------------
// Directive handlers
// -------------------------------------------------------------------------
void CSharpLexerBase::handleDefine()
{
    auto line = collectLine();
    auto sym = symbolFromLine(line);
    if (isActive() && !sym.empty()) symbols_.insert(sym);
}

void CSharpLexerBase::handleUndef()
{
    auto line = collectLine();
    auto sym = symbolFromLine(line);
    if (isActive() && !sym.empty()) symbols_.erase(sym);
}

std::unique_ptr<antlr4::Token> CSharpLexerBase::handleIf()
{
    auto line = collectLine();
    bool outer = isActive();
    bool result = outer && evaluate(line);
    condition_.push_back(result);
    taken_.push_back(result);
    return result ? nullptr : skipFalseBlock();
}

std::unique_ptr<antlr4::Token> CSharpLexerBase::handleElif()
{
    auto line = collectLine();
    bool alreadyTaken = !taken_.empty() ? taken_.back() : false;
    if (!taken_.empty()) taken_.pop_back();
    if (!condition_.empty()) condition_.pop_back();
    bool outer = isActive();
    bool result = !alreadyTaken && outer && evaluate(line);
    condition_.push_back(result);
    taken_.push_back(alreadyTaken || result);
    return result ? nullptr : skipFalseBlock();
}

std::unique_ptr<antlr4::Token> CSharpLexerBase::handleElse()
{
    collectLine();
    bool alreadyTaken = !taken_.empty() ? taken_.back() : false;
    if (!taken_.empty()) taken_.pop_back();
    if (!condition_.empty()) condition_.pop_back();
    bool outer = isActive();
    bool result = !alreadyTaken && outer;
    condition_.push_back(result);
    taken_.push_back(true);
    return result ? nullptr : skipFalseBlock();
}

void CSharpLexerBase::handleEndif()
{
    collectLine();
    if (!condition_.empty()) condition_.pop_back();
    if (!taken_.empty())    taken_.pop_back();
}

// -------------------------------------------------------------------------
// collectLine — drain DIRECTIVE_MODE tokens; tokens owned by lineStorage_
// -------------------------------------------------------------------------
std::vector<antlr4::Token *> CSharpLexerBase::collectLine()
{
    lineStorage_.clear();
    std::vector<antlr4::Token *> result;
    while (true)
    {
        auto t = antlr4::Lexer::nextToken();
        antlr4::Token *raw = t.get();
        if (raw->getChannel() != antlr4::Token::HIDDEN_CHANNEL
            && raw->getChannel() != CSharpLexer::COMMENTS_CHANNEL)
            result.push_back(raw);
        bool done = (raw->getType() == CSharpLexer::DIRECTIVE_NEW_LINE
                  || raw->getType() == antlr4::Token::EOF);
        lineStorage_.push_back(std::move(t));
        if (done) break;
    }
    return result;
}

std::string CSharpLexerBase::symbolFromLine(const std::vector<antlr4::Token *> &line)
{
    for (auto *t : line)
        if (t && t->getType() == CSharpLexer::CONDITIONAL_SYMBOL)
            return t->getText();
    return "";
}

// -------------------------------------------------------------------------
// skipFalseBlock — scan char stream, return SKIPPED_SECTION on HIDDEN channel
// -------------------------------------------------------------------------
std::unique_ptr<antlr4::Token> CSharpLexerBase::skipFalseBlock()
{
    std::string text;
    auto *stream = dynamic_cast<antlr4::CharStream *>(getInputStream());
    int depth = 1;
    bool atLineStart = true;
    size_t startLine = getLine();

    while (true)
    {
        int c = stream->LA(1);
        if (c == antlr4::CharStream::EOF) break;

        if (c == '\r' || c == '\n' || c == 0x85 || c == 0x2028 || c == 0x2029)
        {
            stream->consume();
            text += static_cast<char>(c);
            if (c == '\r' && stream->LA(1) == '\n')
            {
                stream->consume();
                text += '\n';
            }
            atLineStart = true;
            continue;
        }

        if (atLineStart && (c == ' ' || c == '\t'))
        {
            stream->consume();
            text += static_cast<char>(c);
            continue;
        }

        if (atLineStart && c == '#')
        {
            std::string kw = peekKeyword();
            if (kw == "if")
                depth++;
            else if (kw == "endif")
            {
                if (--depth == 0) break;
            }
            else if ((kw == "else" || kw == "elif") && depth == 1)
                break;
        }

        atLineStart = false;
        stream->consume();
        text += static_cast<char>(c);
    }

    auto tok = _factory->create(
        {this, getInputStream()},
        CSharpLexer::SKIPPED_SECTION,
        text,
        antlr4::Token::HIDDEN_CHANNEL,
        -1, -1,
        static_cast<int>(startLine), 0);
    return tok;
}

std::string CSharpLexerBase::peekKeyword()
{
    auto *stream = dynamic_cast<antlr4::CharStream *>(getInputStream());
    int i = 2; // LA(1) is '#'
    while (stream->LA(i) == ' ' || stream->LA(i) == '\t') i++;
    std::string kw;
    int c;
    while ((c = stream->LA(i)) != antlr4::CharStream::EOF && std::isalpha(c))
    {
        kw += static_cast<char>(c);
        i++;
    }
    return kw;
}

// -------------------------------------------------------------------------
// Recursive-descent expression evaluator
// -------------------------------------------------------------------------
bool CSharpLexerBase::evaluate(const std::vector<antlr4::Token *> &tokens)
{
    exprTokens_ = tokens;
    epos_ = 0;
    return parseOr();
}

int CSharpLexerBase::peekType()
{
    if (epos_ < static_cast<int>(exprTokens_.size()))
    {
        int t = exprTokens_[epos_]->getType();
        if (t != CSharpLexer::DIRECTIVE_NEW_LINE && t != antlr4::Token::EOF)
            return t;
    }
    return -1;
}

antlr4::Token *CSharpLexerBase::eConsume() { return exprTokens_[epos_++]; }

bool CSharpLexerBase::parseOr()
{
    bool v = parseAnd();
    while (peekType() == CSharpLexer::OP_OR) { eConsume(); v = parseAnd() || v; }
    return v;
}

bool CSharpLexerBase::parseAnd()
{
    bool v = parseEq();
    while (peekType() == CSharpLexer::OP_AND) { eConsume(); v = parseEq() && v; }
    return v;
}

bool CSharpLexerBase::parseEq()
{
    bool v = parseUnary();
    if (peekType() == CSharpLexer::OP_EQ) { eConsume(); return v == parseUnary(); }
    if (peekType() == CSharpLexer::OP_NE) { eConsume(); return v != parseUnary(); }
    return v;
}

bool CSharpLexerBase::parseUnary()
{
    if (peekType() == CSharpLexer::BANG) { eConsume(); return !parseUnary(); }
    return parsePrimary();
}

bool CSharpLexerBase::parsePrimary()
{
    int t = peekType();
    if (t == CSharpLexer::TRUE)               { eConsume(); return true; }
    if (t == CSharpLexer::FALSE)              { eConsume(); return false; }
    if (t == CSharpLexer::CONDITIONAL_SYMBOL) { return symbols_.count(eConsume()->getText()) > 0; }
    if (t == CSharpLexer::OPEN_PARENS)
    {
        eConsume();
        bool v = parseOr();
        if (peekType() == CSharpLexer::CLOSE_PARENS) eConsume();
        return v;
    }
    return false;
}
